### ISSUE TITLE
[TW-1136] Return useful values to the shell

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#1171]
_Pietro Cerutti on 2009-08-25T14:12:55Z says:_ 

 It would be nice if task returned some useful value to the shell, e.g. 1 if no tasks match the specified criteria. This would help script integration a lot.

The attached patch implements such a modification. Enhancements, feedback, flames welcome :)

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2009-08-25T14:12:55Z
Modified: 2014-02-09T02:36:08Z
```

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2009-08-26T21:43:40Z says:_ 

 Pietro, this is a good idea.  Thank you.

First, I'm moving this feature to 2.0.0, rather than 1.8.2, because 1.8.2 is reserved for bug fixes, and 2.0.0 is what we are actually working on now.

Secondly, I'm rejecting the patch, because it depends on the string length of the output, rather than the number of matching tasks in a query.  The two are loosely related, but I don't want to turn off the diagnostic message to make this work.  I will add a method to Context that returns a matching task count, which will make this work without modifying the output.

That said, I appreciate the good idea, and for illustrating the need to provide internal support for this.  Thanks!

**Comment by taskwarrior** [comment]
_Pietro Cerutti on 2009-08-27T01:57:29Z says:_ 

 Paul,
thanks for taking this into account!

I agree that relying on the length of the returned string is poor design, but please note that I'm not turning off any diagnostic messages. The printing of the "No matches." messages is just deferred until Context.

Anyway, I like your idea of adding a method to return the number of tasks matching a given request from Context, as I was somehow missing it while first rolling out this patch. I will take care of this and provide an updated patch :)

Thanks!

**Comment by taskwarrior** [comment]
_Pietro Cerutti on 2009-08-27T11:04:44Z says:_ 

 Paul,

I have rolled a new patch. It changes quite a bit the logic of the "command" methods. The basic pattern change is
*  std::string handleSomeCommand(void);
+ int         handleSomeCommand(std::string &);

commands return 0 on success or some value to indicate failure and save their output on the string reference they receive.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2009-08-29T09:26:38Z says:_ 

 Pietro,

Thank you for the updated patch.  I studied it closely (hence the delay).  Your changes are surgical and minimal, without any discernible flaws, and with no effect on the test suite.  They also add a very nice feature.  The patch has been applied to the 1.8.2 branch and pushed.  If 1.8.2 is never released, the changes there will be merged to 2.0.0, which is just another way of saying that your patch will be part of the next and all subsequent releases.  Earlier, I said exactly the opposite, and I was wrong - this should be included in the next release.

I verified that your name is listed in the AUTHORS file.

I took the liberty of add a feature.exit.t unit test to prove this new functionality, which by omission is the only "complaint" I have.

Thanks for the good work.

**Comment by pbeckingham** [comment]
[TW-1136_task_return.diff.txt](https://github.com/GothenburgBitFactory/taskwarrior/files/1756294/TW-1136_task_return.diff.txt)
[TW-1136_task_return2.diff.txt](https://github.com/GothenburgBitFactory/taskwarrior/files/1756295/TW-1136_task_return2.diff.txt)
